Ingest writer memory consumption raised TiDB OOM in the million tables of dml and ddl scenario

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
3 million tables and dml
ddl(add column/index, change column type) of concurrency 64
![Image](https://github.com/user-attachments/assets/f9f31bfa-3a6e-4def-8dc4-464a463364ce)

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->
 Release Version: v8.5.0-alpha-410-gd0ea9e50d5
Edition: Community
Git Commit Hash: d0ea9e50d5a9d78269172c95d641b49293087f5d
Git Branch: HEAD
UTC Build Time: 2024-12-21 05:26:21
GoVersion: go1.23.4
Race Enabled: false
Check Table Before Drop: false
Store: unistore

```
2024/12/29 21:10:43.154 +00:00] [INFO] [task_executor.go:143] ["subtask is scheduled away, cancel running"] [task-id=1587] [task-type=backfill] [subtaskID=1587]
...
[2024/12/29 21:10:43.155 +00:00] [INFO] [task_executor.go:188] ["update subtask row count failed"] [task-id=1587] [task-type=backfill] [error="context canceled"] [errorVerbose="context cancel
ed\ngithub.com...
[2024/12/29 21:10:43.156 +00:00] [WARN] [handle.go:228] ["met retryable error"] [task-id=1587] [task-type=backfill] [retry-count=0] [max-retry=30] [error="context canceled"] [errorVerbose="co
ntext canceled...
[2024/12/29 21:10:43.156 +00:00] [ERROR] [task_executor.go:325] ["run subtask failed"] [task-id=1587] [task-type=backfill] [error="context canceled"]
...
[2024/12/29 21:10:43.460 +00:00] [INFO] [task_executor.go:402] ["subtask in running state and is idempotent"] [task-id=1587] [task-type=backfill] [subtask-id=1587]
[2024/12/29 21:10:43.460 +00:00] [INFO] [task_executor.go:418] ["run subtask start"] [task-id=1587] [task-type=backfill] [subtaskID=1587] [step=read-index]
[2024/12/29 21:10:43.460 +00:00] [INFO] [backfilling_read_index.go:98] ["read index executor run subtask"] [category=ddl] ["use cloud"=false]
[2024/12/29 21:10:43.492 +00:00] [ERROR] [backfilling_operators.go:1006] ["import error"] [task-id=1587] [subtask-id=1587] [category=ddl] [error="context canceled"]
[2024/12/29 21:10:43.492 +00:00] [INFO] [task_executor.go:440] ["run subtask failed"] [task-id=1587] [task-type=backfill] [subtaskID=1587] [step=read-index] [takeTime=32.071207ms] [error="con
text canceled"] [errorVerbose="context canceled\ngithub.com/tikv/pd/client/clients/tso.(*Request).waitCtx\n\t/root/go/pkg/mod/github.com/tikv/pd/client@v0.0.0-20241220053006-461b86adc78d/clie
nts/tso/request.go:98\ngithub.com...
[2024/12/29 21:10:43.796 +00:00] [INFO] [task_executor.go:402] ["subtask in running state and is idempotent"] [task-id=1587] [task-type=backfill] [subtask-id=1587]
[2024/12/29 21:10:43.796 +00:00] [INFO] [task_executor.go:418] ["run subtask start"] [task-id=1587] [task-type=backfill] [subtaskID=1587] [step=read-index]
...
```

If the subtask is scheduled away and it comes back immediately, step executor always uses the old, canceled `stepCtx`:

https://github.com/pingcap/tidb/blob/593451549b78143d2b3d6ecd5b7b6ec08032f530/pkg/disttask/framework/taskexecutor/task_executor.go#L142-L148

--- a/pkg/disttask/framework/scheduler/balancer.go
+++ b/pkg/disttask/framework/scheduler/balancer.go
@@ -19,6 +19,7 @@ import (
 	"time"
 
 	"github.com/pingcap/errors"
+	"github.com/pingcap/failpoint"
 	"github.com/pingcap/tidb/pkg/disttask/framework/proto"
 	llog "github.com/pingcap/tidb/pkg/lightning/log"
 	"github.com/pingcap/tidb/pkg/util/intest"
@@ -116,6 +117,9 @@ func (b *balancer) doBalanceSubtasks(ctx context.Context, taskID int64, eligible
 	// managed nodes, subtasks of task might not be balanced.
 	adjustedNodes := filterNodesWithEnoughSlots(b.currUsedSlots, b.slotMgr.getCapacity(),
 		eligibleNodes, subtasks[0].Concurrency)
+	failpoint.Inject("mockNoEnoughSlots", func(_ failpoint.Value) {
+		adjustedNodes = []string{}
+	})
 	if len(adjustedNodes) == 0 {
 		// no node has enough slots to run the subtasks, skip balance and skip
 		// update used slots.

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -144,6 +144,7 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context) {
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
 			e.cancelRunStepWith(nil)
+			failpoint.InjectCall("afterCancelRunningSubtask")
 			return
 		}
 
@@ -307,7 +308,9 @@ func (e *BaseTaskExecutor) Run() {
 		// reset it when we get a subtask
 		checkInterval, noSubtaskCheckCnt = SubtaskCheckInterval, 0
 
-		if e.stepExec != nil && e.stepExec.GetStep() != subtask.Step {
+		if e.stepExec != nil &&
+			(e.stepExec.GetStep() != subtask.Step ||
+				e.stepCtx.Err() != nil) { // Previous step ctx is done, cleanup and use a new one.
 			e.cleanStepExecutor()
 		}
 		if e.stepExec == nil {

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -90,6 +90,7 @@ type BaseTaskExecutor struct {
 		sync.RWMutex
 		// runtimeCancel is used to cancel the Run/Rollback when error occurs.
 		runtimeCancel context.CancelCauseFunc
+		subtaskCancel context.CancelFunc
 	}
 
 	stepExec   execute.StepExecutor
@@ -143,7 +144,7 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context) {
 			e.logger.Info("subtask is scheduled away, cancel running",
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
-			e.Cancel()
+			e.cancelSubtaskCtx()
 			failpoint.InjectCall("afterCancelSubtaskExec")
 			return
 		}
@@ -425,23 +426,28 @@ func (e *BaseTaskExecutor) runSubtask(subtask *proto.Subtask) (resErr error) {
 	logTask := llog.BeginTask(logger, "run subtask")
 	subtaskErr := func() error {
 		e.currSubtaskID.Store(subtask.ID)
+		subtaskCtx, subtaskCancelCtx := context.WithCancel(e.stepCtx)
+		e.mu.Lock()
+		e.mu.subtaskCancel = subtaskCancelCtx
+		e.mu.Unlock()
 
 		var wg util.WaitGroupWrapper
-		checkCtx, checkCancel := context.WithCancel(e.stepCtx)
+		checkCtx, checkCancel := context.WithCancel(subtaskCtx)
 		wg.RunWithLog(func() {
 			e.checkBalanceSubtask(checkCtx)
 		})
 
 		if e.hasRealtimeSummary(e.stepExec) {
 			wg.RunWithLog(func() {
-				e.updateSubtaskSummaryLoop(checkCtx, e.stepCtx, e.stepExec)
+				e.updateSubtaskSummaryLoop(checkCtx, subtaskCtx, e.stepExec)
 			})
 		}
 		defer func() {
 			checkCancel()
+			subtaskCancelCtx()
 			wg.Wait()
 		}()
-		return e.stepExec.RunSubtask(e.stepCtx, subtask)
+		return e.stepExec.RunSubtask(subtaskCtx, subtask)
 	}()
 	failpoint.InjectCall("afterRunSubtask", e, &subtaskErr)
 	logTask.End2(zap.InfoLevel, subtaskErr)
@@ -491,6 +497,14 @@ func (e *BaseTaskExecutor) cancelRunStepWith(cause error) {
 	defer e.mu.Unlock()
 	if e.mu.runtimeCancel != nil {
 		e.mu.runtimeCancel(cause)
+	}
+}
+
+func (e *BaseTaskExecutor) cancelSubtaskCtx() {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	if e.mu.subtaskCancel != nil {
+		e.mu.subtaskCancel()
 	}
 }
 

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -143,7 +143,9 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context, subtaskCtxCa
 			e.logger.Info("subtask is scheduled away, cancel running",
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
-			subtaskCtxCancel()
+			if subtaskCtxCancel != nil {
+				subtaskCtxCancel()
+			}
 			failpoint.InjectCall("afterCancelSubtaskExec")
 			return
 		}

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -90,7 +90,6 @@ type BaseTaskExecutor struct {
 		sync.RWMutex
 		// runtimeCancel is used to cancel the Run/Rollback when error occurs.
 		runtimeCancel context.CancelCauseFunc
-		subtaskCancel context.CancelFunc
 	}
 
 	stepExec   execute.StepExecutor
@@ -123,7 +122,7 @@ func NewBaseTaskExecutor(ctx context.Context, task *proto.Task, param Param) *Ba
 //     `pending` state, to make sure subtasks can be balanced later when node scale out.
 //   - If current running subtask are scheduled away from this node, i.e. this node
 //     is taken as down, cancel running.
-func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context) {
+func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context, subtaskCancelCtx context.CancelFunc) {
 	ticker := time.NewTicker(checkBalanceSubtaskInterval)
 	defer ticker.Stop()
 	for {
@@ -144,7 +143,7 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context) {
 			e.logger.Info("subtask is scheduled away, cancel running",
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
-			e.cancelSubtaskCtx()
+			subtaskCancelCtx()
 			failpoint.InjectCall("afterCancelSubtaskExec")
 			return
 		}
@@ -427,14 +426,11 @@ func (e *BaseTaskExecutor) runSubtask(subtask *proto.Subtask) (resErr error) {
 	subtaskErr := func() error {
 		e.currSubtaskID.Store(subtask.ID)
 		subtaskCtx, subtaskCancelCtx := context.WithCancel(e.stepCtx)
-		e.mu.Lock()
-		e.mu.subtaskCancel = subtaskCancelCtx
-		e.mu.Unlock()
 
 		var wg util.WaitGroupWrapper
 		checkCtx, checkCancel := context.WithCancel(subtaskCtx)
 		wg.RunWithLog(func() {
-			e.checkBalanceSubtask(checkCtx)
+			e.checkBalanceSubtask(checkCtx, subtaskCancelCtx)
 		})
 
 		if e.hasRealtimeSummary(e.stepExec) {
@@ -497,14 +493,6 @@ func (e *BaseTaskExecutor) cancelRunStepWith(cause error) {
 	defer e.mu.Unlock()
 	if e.mu.runtimeCancel != nil {
 		e.mu.runtimeCancel(cause)
-	}
-}
-
-func (e *BaseTaskExecutor) cancelSubtaskCtx() {
-	e.mu.Lock()
-	defer e.mu.Unlock()
-	if e.mu.subtaskCancel != nil {
-		e.mu.subtaskCancel()
 	}
 }
 

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -122,7 +122,7 @@ func NewBaseTaskExecutor(ctx context.Context, task *proto.Task, param Param) *Ba
 //     `pending` state, to make sure subtasks can be balanced later when node scale out.
 //   - If current running subtask are scheduled away from this node, i.e. this node
 //     is taken as down, cancel running.
-func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context, subtaskCancelCtx context.CancelFunc) {
+func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context, subtaskCtxCancel context.CancelFunc) {
 	ticker := time.NewTicker(checkBalanceSubtaskInterval)
 	defer ticker.Stop()
 	for {
@@ -143,7 +143,7 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context, subtaskCance
 			e.logger.Info("subtask is scheduled away, cancel running",
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
-			subtaskCancelCtx()
+			subtaskCtxCancel()
 			failpoint.InjectCall("afterCancelSubtaskExec")
 			return
 		}
@@ -425,12 +425,12 @@ func (e *BaseTaskExecutor) runSubtask(subtask *proto.Subtask) (resErr error) {
 	logTask := llog.BeginTask(logger, "run subtask")
 	subtaskErr := func() error {
 		e.currSubtaskID.Store(subtask.ID)
-		subtaskCtx, subtaskCancelCtx := context.WithCancel(e.stepCtx)
+		subtaskCtx, subtaskCtxCancel := context.WithCancel(e.stepCtx)
 
 		var wg util.WaitGroupWrapper
 		checkCtx, checkCancel := context.WithCancel(subtaskCtx)
 		wg.RunWithLog(func() {
-			e.checkBalanceSubtask(checkCtx, subtaskCancelCtx)
+			e.checkBalanceSubtask(checkCtx, subtaskCtxCancel)
 		})
 
 		if e.hasRealtimeSummary(e.stepExec) {
@@ -440,8 +440,8 @@ func (e *BaseTaskExecutor) runSubtask(subtask *proto.Subtask) (resErr error) {
 		}
 		defer func() {
 			checkCancel()
-			subtaskCancelCtx()
 			wg.Wait()
+			subtaskCtxCancel()
 		}()
 		return e.stepExec.RunSubtask(subtaskCtx, subtask)
 	}()

--- a/pkg/disttask/framework/taskexecutor/task_executor.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor.go
@@ -143,8 +143,8 @@ func (e *BaseTaskExecutor) checkBalanceSubtask(ctx context.Context) {
 			e.logger.Info("subtask is scheduled away, cancel running",
 				zap.Int64("subtaskID", e.currSubtaskID.Load()))
 			// cancels runStep, but leave the subtask state unchanged.
-			e.cancelRunStepWith(nil)
-			failpoint.InjectCall("afterCancelRunningSubtask")
+			e.Cancel()
+			failpoint.InjectCall("afterCancelSubtaskExec")
 			return
 		}
 
@@ -308,9 +308,7 @@ func (e *BaseTaskExecutor) Run() {
 		// reset it when we get a subtask
 		checkInterval, noSubtaskCheckCnt = SubtaskCheckInterval, 0
 
-		if e.stepExec != nil &&
-			(e.stepExec.GetStep() != subtask.Step ||
-				e.stepCtx.Err() != nil) { // Previous step ctx is done, cleanup and use a new one.
+		if e.stepExec != nil && e.stepExec.GetStep() != subtask.Step {
 			e.cleanStepExecutor()
 		}
 		if e.stepExec == nil {
@@ -319,6 +317,12 @@ func (e *BaseTaskExecutor) Run() {
 					zap.String("step", proto.Step2Str(task.Type, task.Step)), zap.Error(err2))
 				continue
 			}
+		}
+		if err := e.stepCtx.Err(); err != nil {
+			e.logger.Error("step executor context is done, the task should have been reverted",
+				zap.String("step", proto.Step2Str(task.Type, task.Step)),
+				zap.Error(err))
+			continue
 		}
 		err = e.runSubtask(subtask)
 		if err != nil {

--- a/pkg/disttask/framework/taskexecutor/task_executor_test.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor_test.go
@@ -326,6 +326,19 @@ func TestTaskExecutorRun(t *testing.T) {
 			<-ctx.Done()
 			return ctx.Err()
 		})
+		e.taskExecExt.EXPECT().IsRetryableError(gomock.Any()).Return(true)
+		// keep running next subtask
+		nextSubtask := &proto.Subtask{SubtaskBase: proto.SubtaskBase{
+			ID: 2, Type: e.task1.Type, Step: proto.StepOne, State: proto.SubtaskStatePending, ExecID: "id"}}
+		e.taskTable.EXPECT().GetTaskByID(gomock.Any(), e.task1.ID).Return(e.task1, nil)
+		e.taskTable.EXPECT().GetFirstSubtaskInStates(gomock.Any(), "id", e.task1.ID, proto.StepOne,
+			unfinishedNormalSubtaskStates...).Return(nextSubtask, nil)
+		e.stepExecutor.EXPECT().GetStep().Return(proto.StepOne)
+		e.taskTable.EXPECT().StartSubtask(gomock.Any(), nextSubtask.ID, "id").Return(nil)
+		e.stepExecutor.EXPECT().RunSubtask(gomock.Any(), nextSubtask).Return(nil)
+		e.taskTable.EXPECT().FinishSubtask(gomock.Any(), "id", nextSubtask.ID, gomock.Any()).Return(nil)
+		// exit
+		e.taskTable.EXPECT().GetTaskByID(gomock.Any(), e.task1.ID).Return(e.succeedTask1, nil)
 		e.stepExecutor.EXPECT().Cleanup(gomock.Any()).Return(nil)
 		e.taskExecutor.Run()
 		require.True(t, e.ctrl.Satisfied())
@@ -886,7 +899,7 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		mockSubtaskTable.EXPECT().GetSubtasksByExecIDAndStepAndStates(gomock.Any(), "tidb1",
 			task.ID, task.Step, proto.SubtaskStateRunning).Return([]*proto.Subtask{}, nil)
 		runCtx, cancel := context.WithCancel(ctx)
-		taskExecutor.cancel = cancel
+		taskExecutor.mu.subtaskCancel = cancel
 		require.NoError(t, runCtx.Err())
 		taskExecutor.checkBalanceSubtask(ctx)
 		require.ErrorIs(t, runCtx.Err(), context.Canceled)

--- a/pkg/disttask/framework/taskexecutor/task_executor_test.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor_test.go
@@ -897,8 +897,8 @@ func TestCheckBalanceSubtask(t *testing.T) {
 			task.ID, task.Step, proto.SubtaskStateRunning).Return(nil, errors.New("error"))
 		mockSubtaskTable.EXPECT().GetSubtasksByExecIDAndStepAndStates(gomock.Any(), "tidb1",
 			task.ID, task.Step, proto.SubtaskStateRunning).Return([]*proto.Subtask{}, nil)
-		runCtx, cancelCause := context.WithCancelCause(ctx)
-		taskExecutor.mu.runtimeCancel = cancelCause
+		runCtx, cancel := context.WithCancel(ctx)
+		taskExecutor.cancel = cancel
 		require.NoError(t, runCtx.Err())
 		taskExecutor.checkBalanceSubtask(ctx)
 		require.ErrorIs(t, runCtx.Err(), context.Canceled)

--- a/pkg/disttask/framework/taskexecutor/task_executor_test.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor_test.go
@@ -315,7 +315,7 @@ func TestTaskExecutorRun(t *testing.T) {
 		// mock for checkBalanceSubtask, returns empty subtask list
 		e.taskTable.EXPECT().GetSubtasksByExecIDAndStepAndStates(gomock.Any(), "id",
 			e.task1.ID, proto.StepOne, proto.SubtaskStateRunning).Return([]*proto.Subtask{}, nil)
-		// this subtask is scheduled awsy during running
+		// this subtask is scheduled away during running
 		e.taskTable.EXPECT().GetTaskByID(gomock.Any(), e.task1.ID).Return(e.task1, nil)
 		e.taskTable.EXPECT().GetFirstSubtaskInStates(gomock.Any(), "id", e.task1.ID, proto.StepOne,
 			unfinishedNormalSubtaskStates...).Return(e.pendingSubtask1, nil)
@@ -326,18 +326,6 @@ func TestTaskExecutorRun(t *testing.T) {
 			<-ctx.Done()
 			return ctx.Err()
 		})
-		// keep running next subtask
-		nextSubtask := &proto.Subtask{SubtaskBase: proto.SubtaskBase{
-			ID: 2, Type: e.task1.Type, Step: proto.StepOne, State: proto.SubtaskStatePending, ExecID: "id"}}
-		e.taskTable.EXPECT().GetTaskByID(gomock.Any(), e.task1.ID).Return(e.task1, nil)
-		e.taskTable.EXPECT().GetFirstSubtaskInStates(gomock.Any(), "id", e.task1.ID, proto.StepOne,
-			unfinishedNormalSubtaskStates...).Return(nextSubtask, nil)
-		e.stepExecutor.EXPECT().GetStep().Return(proto.StepOne)
-		e.taskTable.EXPECT().StartSubtask(gomock.Any(), nextSubtask.ID, "id").Return(nil)
-		e.stepExecutor.EXPECT().RunSubtask(gomock.Any(), nextSubtask).Return(nil)
-		e.taskTable.EXPECT().FinishSubtask(gomock.Any(), "id", nextSubtask.ID, gomock.Any()).Return(nil)
-		// exit
-		e.taskTable.EXPECT().GetTaskByID(gomock.Any(), e.task1.ID).Return(e.succeedTask1, nil)
 		e.stepExecutor.EXPECT().Cleanup(gomock.Any()).Return(nil)
 		e.taskExecutor.Run()
 		require.True(t, e.ctrl.Satisfied())

--- a/pkg/disttask/framework/taskexecutor/task_executor_test.go
+++ b/pkg/disttask/framework/taskexecutor/task_executor_test.go
@@ -890,7 +890,7 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		// context canceled
 		canceledCtx, cancel := context.WithCancel(ctx)
 		cancel()
-		taskExecutor.checkBalanceSubtask(canceledCtx)
+		taskExecutor.checkBalanceSubtask(canceledCtx, nil)
 	})
 
 	t.Run("subtask scheduled away", func(t *testing.T) {
@@ -899,9 +899,8 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		mockSubtaskTable.EXPECT().GetSubtasksByExecIDAndStepAndStates(gomock.Any(), "tidb1",
 			task.ID, task.Step, proto.SubtaskStateRunning).Return([]*proto.Subtask{}, nil)
 		runCtx, cancel := context.WithCancel(ctx)
-		taskExecutor.mu.subtaskCancel = cancel
 		require.NoError(t, runCtx.Err())
-		taskExecutor.checkBalanceSubtask(ctx)
+		taskExecutor.checkBalanceSubtask(ctx, cancel)
 		require.ErrorIs(t, runCtx.Err(), context.Canceled)
 		require.True(t, ctrl.Satisfied())
 	})
@@ -914,7 +913,7 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		mockExtension.EXPECT().IsIdempotent(subtasks[0]).Return(false)
 		mockSubtaskTable.EXPECT().UpdateSubtaskStateAndError(gomock.Any(), "tidb1",
 			subtasks[0].ID, proto.SubtaskStateFailed, ErrNonIdempotentSubtask).Return(nil)
-		taskExecutor.checkBalanceSubtask(ctx)
+		taskExecutor.checkBalanceSubtask(ctx, nil)
 		require.True(t, ctrl.Satisfied())
 
 		// if we failed to change state of non-idempotent subtask, will retry
@@ -931,7 +930,7 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		mockExtension.EXPECT().IsIdempotent(subtasks[0]).Return(false)
 		mockSubtaskTable.EXPECT().UpdateSubtaskStateAndError(gomock.Any(), "tidb1",
 			subtasks[0].ID, proto.SubtaskStateFailed, ErrNonIdempotentSubtask).Return(nil)
-		taskExecutor.checkBalanceSubtask(ctx)
+		taskExecutor.checkBalanceSubtask(ctx, nil)
 		require.True(t, ctrl.Satisfied())
 	})
 
@@ -946,7 +945,7 @@ func TestCheckBalanceSubtask(t *testing.T) {
 		// used to break the loop
 		mockSubtaskTable.EXPECT().GetSubtasksByExecIDAndStepAndStates(gomock.Any(), "tidb1",
 			task.ID, task.Step, proto.SubtaskStateRunning).Return(nil, nil)
-		taskExecutor.checkBalanceSubtask(ctx)
+		taskExecutor.checkBalanceSubtask(ctx, nil)
 		require.True(t, ctrl.Satisfied())
 	})
 }

--- a/tests/realtikvtest/addindextest1/disttask_test.go
+++ b/tests/realtikvtest/addindextest1/disttask_test.go
@@ -391,3 +391,48 @@ func TestAddIndexDistLockAcquireFailed(t *testing.T) {
 	testfailpoint.Enable(t, "github.com/pingcap/tidb/pkg/owner/mockAcquireDistLockFailed", "1*return(true)")
 	tk.MustExec("alter table t add index idx(b);")
 }
+
+func TestAddIndexScheduleAway(t *testing.T) {
+	store := realtikvtest.CreateMockStoreAndSetup(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("use test")
+	tk.MustExec("set global tidb_enable_dist_task = on;")
+	t.Cleanup(func() {
+		tk.MustExec("set global tidb_enable_dist_task = off;")
+	})
+	tk.MustExec("create table t (a int, b int);")
+	tk.MustExec("insert into t values (1, 1);")
+
+	var jobID atomic.Int64
+	// Acquire the job ID.
+	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/ddl/beforeRunOneJobStep", func(job *model.Job) {
+		if job.Type == model.ActionAddIndex {
+			jobID.Store(job.ID)
+		}
+	})
+	// Do not balance subtasks automatically.
+	testfailpoint.Enable(t, "github.com/pingcap/tidb/pkg/disttask/framework/scheduler/mockNoEnoughSlots", "return")
+	afterCancel := make(chan struct{})
+	// Capture the cancel operation from checkBalanceLoop.
+	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor/afterCancelRunningSubtask", func() {
+		close(afterCancel)
+	})
+	var once sync.Once
+	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/ddl/mockDMLExecutionAddIndexSubTaskFinish", func() {
+		once.Do(func() {
+			tk1 := testkit.NewTestKit(t, store)
+			tk1.MustExec("use test")
+			updateExecID := fmt.Sprintf(`
+				update mysql.tidb_background_subtask set exec_id = 'other' where task_key in 
+					(select id from mysql.tidb_global_task where task_key like '%%%d')`, jobID.Load())
+			tk1.MustExec(updateExecID)
+			<-afterCancel
+			updateExecID = fmt.Sprintf(`
+				update mysql.tidb_background_subtask set exec_id = ':4000' where task_key in 
+					(select id from mysql.tidb_global_task where task_key like '%%%d')`, jobID.Load())
+			tk1.MustExec(updateExecID)
+		})
+	})
+	tk.MustExec("alter table t add index idx(b);")
+	require.NotEqual(t, int64(0), jobID.Load())
+}

--- a/tests/realtikvtest/addindextest1/disttask_test.go
+++ b/tests/realtikvtest/addindextest1/disttask_test.go
@@ -414,7 +414,7 @@ func TestAddIndexScheduleAway(t *testing.T) {
 	testfailpoint.Enable(t, "github.com/pingcap/tidb/pkg/disttask/framework/scheduler/mockNoEnoughSlots", "return")
 	afterCancel := make(chan struct{})
 	// Capture the cancel operation from checkBalanceLoop.
-	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor/afterCancelRunningSubtask", func() {
+	testfailpoint.EnableCall(t, "github.com/pingcap/tidb/pkg/disttask/framework/taskexecutor/afterCancelSubtaskExec", func() {
 		close(afterCancel)
 	})
 	var once sync.Once

